server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1078,7 +1083,40 @@ private void fillSegmentInfo(SegmentReader segmentReader, boolean search, Map<St
             logger.trace(() -> "failed to get size for [" + info.info.name + "]", e);
         }
         segment.segmentSort = info.info.getIndexSort();
-        segment.attributes = info.info.getAttributes();
+        segment.attributes = new HashMap<>();
+        segment.attributes.putAll(info.info.getAttributes());
+        Codec codec = info.info.getCodec();
+        Map<String, List<String>> knnFormats = null;
+        if (codec instanceof PerFieldMapperCodec) {
+            try {
+                FieldInfos fieldInfos = segmentReader.getFieldInfos();
+                if (fieldInfos.hasVectorValues()) {
+                    for (FieldInfo fieldInfo : fieldInfos) {
+                        String name = fieldInfo.getName();
+                        if (fieldInfo.hasVectorValues()) {
+                            KnnVectorsFormat knnVectorsFormatForField = ((PerFieldMapperCodec) codec).getKnnVectorsFormatForField(name);
+                            if (knnFormats == null) {
+                                knnFormats = new HashMap<>();
+                            }
+                            knnFormats.compute(knnVectorsFormatForField.getName(), (s, a) -> {
+                                if (a == null) {
+                                    a = new ArrayList<>();
+                                }
+                                a.add(name);
+                                return a;
+                            });
+                        }
+                    }
+                }
+            } catch (AlreadyClosedException ace) {
```

Comment:
I know we have been trying to figure a nice way without adding new API parameters here.

I did double check, and the ONLY thing reading segments in this manner is the segments API. So, adding a parameter to `getSegments` that gets passed down here isn't that big of a deal IMO.

The good thing is the engine doesn't keep this information around on heap, its just read for each shard for that API.

My main concern here when there are many thousands of fields (while only one field has vectors) and hundreds of segments. The overall cost may still be dominated by other factors and this will have negligible impact, I just don't know.

We are indeed adding a loop over (potentially) thousands of items, while 99.999% of them will be noops, we are still looping over them, where as before we were not.

Is there a simple way to benchmark this performance impact? If its very small when we have 10s of thousands of fields (many folks run with this many).

Another option is to not benchmark, and just have a parameter that defaults to false, something like "gatherVectorFieldAttributes" or something.

Reply:
I will add also do some benchmarking, which makes sense anyway.
regarding adding a parameter to the `getSegments` API, I think we need to keep also the no param method(s) for the sake of backward compatibility.

Reply:
> I think we need to keep also the no param method(s) for the sake of backward compatibility.

No doubt, we will just default to `false`. It would be an optional URL parameter

Reply:
the following IT checks that the `segments` call time is not proportional to the number of vector fields in the index.
```java
public void testManyIndicesSegmentsWithVectorsIncluded() {
        String indexName = "test-vectors";
        createIndex(indexName);
        ensureGreen(indexName);

        int numVectorFields = randomIntBetween(100, 1000);
        for (int i = 0; i < numVectorFields; i++) {
            String vectorField = "embedding_" + i;
            addMapping(indexName, vectorField);
            addVectors(indexName, vectorField, 1024, 1000);
        }

        long start = System.currentTimeMillis();
        IndicesSegmentResponse response = indicesAdmin().prepareSegments(indexName).includeVectorFormatInfo(false).get();
        long end = System.currentTimeMillis();
        long timeMS = end - start;

        assertNoFailures(response);
        IndexSegments indexSegments = response.getIndices().get(indexName);
        assertNotNull(indexSegments);
        IndexShardSegments shardSegments = indexSegments.getShards().get(0);
        assertNotNull(shardSegments);

        ShardSegments shard = shardSegments.shards()[0];
        for (Segment segment : shard.getSegments()) {
            assertThat(segment.getAttributes().keySet(), not(hasItem(endsWith("VectorsFormat"))));
        }

        indicesAdmin().prepareClearCache(indexName).get();

        long startIncluded = System.currentTimeMillis();
        IndicesSegmentResponse responseIncluded = indicesAdmin().prepareSegments(indexName).includeVectorFormatInfo(true).get();
        long endIncluded = System.currentTimeMillis();
        long includedTimeMS = endIncluded - startIncluded;

        assertNoFailures(responseIncluded);
        IndexSegments indexSegmentsIncluded = responseIncluded.getIndices().get(indexName);
        assertNotNull(indexSegmentsIncluded);
        IndexShardSegments shardSegmentsIncluded = indexSegmentsIncluded.getShards().get(0);
        assertNotNull(shardSegmentsIncluded);

        shard = shardSegmentsIncluded.shards()[0];
        for (Segment segment : shard.getSegments()) {
            assertThat(segment.getAttributes().keySet(), hasItem(endsWith("VectorsFormat")));
        }

        assertThat(includedTimeMS, lessThan(timeMS * numVectorFields));
    }
```
it always passed in a 100 runs sample.

I've tried turning it into a micro benchmark but it requires way too many dependencies to be wired and, imho, it doesn't make much sense to add it to `rally`, at the same time, it's a bit long/expensive to run (2-5mins depending on randomization) as a pure integration test.